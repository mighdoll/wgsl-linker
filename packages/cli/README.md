## wgsl-link 

**wgsl-link** is a tool for preprocessing and linking multiple WGSL shader modules into a single WGSL module from the command line.

See [wgsl-linker](https://github.com/mighdoll/wgsl-linker) 
for more details of wgsl syntax extensions
like `#import` and `#export`.

### Usage
```
wgsl-link <rootWgsl> [libraryWgsl...]
```

Merges the rootWgsl file with any imports 
found in the libraryWgsl files and outputs a bundled wgsl to stdout.

### Options 
`--define name=value` &ensp; 
defines variables for conditional compilation
and template preocessing. Mutliple name=value pairs can be specified.

`--base-dir <path element>` &ensp;
sets the base directory for resolving relative paths.



