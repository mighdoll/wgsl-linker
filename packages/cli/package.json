{
  "name": "wgsl-link",
  "version": "0.4",
  "type": "commonjs",
  "scripts": {
    "prepublishOnly": "run-s build",
    "build": "esbuild --bundle --platform=node --outfile=bin/wgsl-link src/main.ts"
  },
  "repository": "https://github.com/mighdoll/wgsl-linker/tree/main/packages/cli",
  "homepage": "https://github.com/mighdoll/wgsl-linker/tree/main/packages/cli#readme",
  "files": [
    "bin"
  ],
  "bin": "bin/wgsl-link",
  "dependencies": {
    "@types/diff": "^5.0.9",
    "diff": "^5.2.0",
    "wgsl-linker": "workspace:*",
    "yargs": "^17.7.2"
  },
  "devDependencies": {
    "@esbuild-plugins/tsconfig-paths": "^0.1.2",
    "@types/node": "^20.11.30",
    "@types/yargs": "^17.0.32",
    "esbuild": "^0.20.2",
    "npm-run-all": "^4.1.5",
    "tsx": "^4.7.1",
    "vitest": "^1.4.0"
  }
}
