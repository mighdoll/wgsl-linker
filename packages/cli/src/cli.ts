import yargs from "yargs";
import fs from "fs";
import { ModuleRegistry, normalize } from "wgsl-linker";
import { createTwoFilesPatch } from "diff";
import { TypeRefElem } from "../../linker/src/AbstractElems.js";

type CliArgs = ReturnType<typeof parseArgs>;
let argv: CliArgs;

export async function cli(rawArgs: string[]): Promise<void> {
  argv = parseArgs(rawArgs);
  const files = argv.files as string[];
  argv.separately ? linkSeparately(files) : linkNormally(files);
}

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
function parseArgs(args: string[]) {
  return yargs(args)
    .command("$0 <files...>", "root wgsl file followed by any library wgsl files")
    .option("define", {
      type: "array",
      describe: "definitions for preprocessor and linking",
    })
    .option("baseDir", {
      requiresArg: true,
      type: "string",
      describe: "rm common prefix from file paths",
    })
    .option("separately", {
      type: "boolean",
      default: false,
      hidden: true,
      describe: "link each file separately (for parser testing)",
    })
    .option("details", {
      type: "boolean",
      default: false,
      hidden: true,
      describe: "show details about parsed files",
    })
    .option("diff", {
      type: "boolean",
      default: false,
      hidden: true,
      describe: "show comparison with src file",
    })
    .option("emit", {
      type: "boolean",
      default: true,
      hidden: true,
      describe: "emit linked result",
    })
    .help()
    .parseSync();
}

function linkNormally(paths: string[]): void {
  const pathAndTexts = paths.map((f) => {
    const text = fs.readFileSync(f, { encoding: "utf8" });
    const basedPath = normalize(rmBaseDirPrefix(f));
    return [basedPath, text];
  });
  const wgsl = Object.fromEntries(pathAndTexts);
  const registry = new ModuleRegistry({ wgsl, conditions: externalDefines() });
  const [srcPath, srcText] = pathAndTexts[0];
  doLink(srcPath, registry, srcText);
}

function linkSeparately(paths: string[]): void {
  paths.forEach((f) => {
    const srcText = fs.readFileSync(f, { encoding: "utf8" });
    const basedPath = normalize(rmBaseDirPrefix(f));
    const registry = new ModuleRegistry({ wgsl: { [basedPath]: srcText } });
    doLink(basedPath, registry, srcText);
  });
}

function doLink(
  srcPath: string,
  registry: ModuleRegistry,
  origWgsl: string
): void {
  const linked = registry.link(srcPath, externalDefines());
  argv.emit && console.log(linked);
  argv.diff && printDiff(srcPath, origWgsl, linked);
  argv.details && printDetails(srcPath, registry);
}

function externalDefines(): Record<string, string> {
  if (!argv.define) return {};
  const pairs = argv.define.map((d) => d.toString().split("="));

  const badPair = pairs.find((p) => p.length !== 2);
  if (badPair) {
    console.error("invalid define", badPair);
    return {};
  }

  const withParsedValues = pairs.map(([k, v]) => [k, parseDefineValue(v)]);
  return Object.fromEntries(withParsedValues);
}

function parseDefineValue(value: string): string | number | boolean {
  const v = value.toLowerCase();
  if (v === "true") return true;
  if (v === "false") return false;
  if (value === "NaN") return NaN;
  const n = Number.parseFloat(value);
  if (!Number.isNaN(n)) return n;
  return value;
}

function printDiff(modulePath: string, src: string, linked: string): void {
  if (src !== linked) {
    const patch = createTwoFilesPatch(modulePath, "linked", src, linked);
    console.log(patch);
  } else {
    console.log(`${modulePath}: linked version matches original source`);
  }
}

function printDetails(modulePath: string, registry: ModuleRegistry): void {
  console.log(modulePath, ":");
  const m = registry.findModule(modulePath)!;
  m.fns.forEach((f) => {
    console.log(`  fn ${f.name}`);
    const calls = f.calls.map((c) => c.name).join("  ");
    console.log(`    calls: ${calls}`);
    printTypeRefs(f);
  });
  m.vars.forEach((v) => {
    console.log(`  var ${v.name}`);
    printTypeRefs(v);
  });
  m.structs.forEach((s) => {
    console.log(`  struct ${s.name}`);
    const members = (s.members ?? []).map((m) => m.name).join("  ");
    console.log(`    members: ${members}`);
    printTypeRefs(s);
  });
  console.log();
}

function printTypeRefs(hasTypeRefs: { typeRefs: TypeRefElem[] }): void {
  const typeRefs = hasTypeRefs.typeRefs.map((t) => t.name).join("  ");
  console.log(`    typeRefs: ${typeRefs}`);
}

function rmBaseDirPrefix(path: string): string {
  const baseDir = argv.baseDir;
  if (baseDir) {
    const found = path.indexOf(baseDir);
    if (found !== -1) {
      return path.slice(found + baseDir.length);
    }
  }
  return path;
}
