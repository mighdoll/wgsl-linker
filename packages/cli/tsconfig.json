{
  "compilerOptions": {
    "target": "ESNext",
    "lib": [ "ESNext"],
    "types": ["node"],
    "module": "ESNext",
    "moduleResolution": "Node",
    "esModuleInterop": false,
    "allowSyntheticDefaultImports": true,
    "skipLibCheck": true,
    "allowJs": false,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "importsNotUsedAsValues": "remove",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "paths": {
      "wgsl-linker": ["../linker/src"],
      "mini-parse": ["../mini-parse/src"],
    }
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules"],
}

