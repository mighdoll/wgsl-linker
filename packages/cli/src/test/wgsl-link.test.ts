import { expect, test, vi } from "vitest";
import { cli } from "../cli.js";

test("simple link", async () => {
  const logged = await cliLine(
    `./src/test/wgsl/main.wgsl 
       ./src/test/wgsl/util.wgsl`
  );
  expect(logged).toMatchSnapshot();
});

test("link with definition", async () => {
  const logged = await cliLine(
    `./src/test/wgsl/main.wgsl 
       ./src/test/wgsl/util.wgsl
       --define EXTRA=true`
  );
  expect(logged).to.include("fn extra()");
});

async function cliLine(argsLine: string): Promise<string> {
  return await withConsoleSpy(() => cli(argsLine.split(/\s+/)));
}

async function withConsoleSpy(fn: () => Promise<void>): Promise<string> {
  const consoleSpy = vi.spyOn(console, "log").mockImplementation(() => {});

  await fn();
  const result = consoleSpy.mock.calls[0].join("");
  vi.resetAllMocks();
  return result;
}
