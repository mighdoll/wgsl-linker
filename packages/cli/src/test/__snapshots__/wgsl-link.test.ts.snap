// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`simple link 1`] = `
"fn main() {
  foo();
}

fn foo() {
  // fooImpl
}"
`;
